File: Jim_v5_5/Matlab_Programs/Generate_Single_Channel_Traces.m
clear
%% 1) Select the input tiff file and Create folder for results
[jimPath,~,~] = fileparts(matlab.desktop.editor.getActiveFilename);%Find the location of this script (should be in Jim\Matlab_Programs)
JIM = [fileparts(jimPath),'\Jim_Programs\'];%Convert to the file path for the C++ Jim Programs
[fileName,pathName] = uigetfile('*','Select the Image file');%Open the Dialog box to select the initial file to analyze

overlayColour1 = [1, 0, 0];
overlayColour2 = [0, 1, 0];
overlayColour3 = [0, 0, 1];

completeName = [pathName,fileName];
[~,name,~] = fileparts(completeName);%get the name of the tiff image excluding the .tiff extension
workingDir = [pathName,name];
[~,name,~] = fileparts(workingDir);%also remove the .ome if it exists or any other full stops
workingDir = [pathName,name,'\'];

if ~exist(workingDir, 'dir')
   mkdir(workingDir)%make a subfolder with that name
end
%% 2) Calculate Drifts
iterations = 3;

alignStartFrame = 1;
alignEndFrame = 5;

cmd = [JIM,'Align_Channels.exe "',workingDir,'Aligned" "',completeName,'" -Start ',num2str(alignStartFrame),' -End ',num2str(alignEndFrame),' -Iterations ',num2str(iterations)];%Run the Align_Channels program with the selected image stack as the input and save the results to the results folder with the Aligned prefix
system(cmd);

figure('Name','Before Drift Correction') %Display the initial mean that has no drift correction. This is equivilent to the z projection if the stack in ImageJ
channel1Im = rescale(imread([workingDir,'Aligned_initial_mean.tiff']));
imshow(channel1Im);
truesize([900 900]);

figure('Name','After Drift Correction')%Display the final mean drift corrected mean. 
channel1Im = rescale(imread([workingDir,'Aligned_final_mean.tiff']));
imshow(channel1Im);
truesize([900 900]);

drifts = csvread([workingDir,'Aligned_Drifts.csv'],1);%Read in drifts to see waht the max the image has shifted by
disp(['Maximum drift is ', num2str(max(max(abs(drifts))))]);
%% 3) Make a SubAverage of Frames for Detection 
useMaxProjection = false;

detectionStartFrame = 1;
detectionEndFrame = 25;

maxProjectionString = '';
if useMaxProjection
    maxProjectionString = ' -MaxProjection';
end

cmd = [JIM,'Mean_of_Frames.exe NULL "',workingDir,'Aligned_Drifts.csv" "',workingDir,'Aligned" "',completeName,'" -Start ',num2str(detectionStartFrame),' -End ',num2str(detectionEndFrame),maxProjectionString];
system(cmd);

figure('Name','Sub-Average to use for detection')%Display the mean of the substack that will be used for particle detection
channel1Im = rescale(imread([workingDir,'Aligned_Partial_Mean.tiff']));
imshow(channel1Im);
truesize([900 900]);

%% 4) Detect Particles
% User Defined Parameters 
%Thresholding
cutoff=0.85; % The curoff for the initial thresholding

%Filtering
left = 10;% Excluded particles closer to the edge than this. Make sure this value is larger than the maximum drift. 25 works well in most cases
right = 10;% Excluded particles closer to the edge than this. Make sure this value is larger than the maximum drift. 25 works well in most cases
top = 10;% Excluded particles closer to the edge than this. Make sure this value is larger than the maximum drift. 25 works well in most cases
bottom = 10;% Excluded particles closer to the edge than this. Make sure this value is larger than the maximum drift. 25 works well in most cases


minCount = 10; % Minimum number of pixels in a ROI to be counted as a particle. Use this to exclude speckles of background
maxCount=100; % Maximum number of pixels in a ROI to be counted as a particle. Use this to exclude aggregates

minEccentricity = -0.1; % Eccentricity of best fit ellipse goes from 0 to 1 - 0=Perfect Circle, 1 = Line. Use the Minimum to exclude round objects. Set it to any negative number to allow all round objects
maxEccentricity = 1.1;  % Use the maximum to exclude long, thin objects. Set it to a value above 1 to include long, thin objects  

minLength = 0; % Minimum number of pixels for the major axis of the best fit ellipse
maxLength = 100000; % Maximum number of pixels for the major axis of the best fit ellipse

maxDistFromLinear = 10000000; % Maximum distance that a pixel can diviate from the major axis.


displayMin = 0; % This just adjusts the contrast in the displayed image. It does NOT effect detection
displayMax = 2; % This just adjusts the contrast in the displayed image. It does NOT effect detection
% Detection Program

cmd = [JIM,'Detect_Particles.exe "',workingDir,'Aligned_Partial_Mean.tiff" "',workingDir,'Detected" -BinarizeCutoff ', num2str(cutoff),' -minLength ',num2str(minLength),' -maxLength ',num2str(maxLength),' -minCount ',num2str(minCount),' -maxCount ',num2str(maxCount),' -minEccentricity ',num2str(minEccentricity),' -maxEccentricity ',num2str(maxEccentricity),' -left ',num2str(left),' -right ',num2str(right),' -top ',num2str(top),' -bottom ',num2str(bottom),' -maxDistFromLinear ',num2str(maxDistFromLinear)]; % Run the program Find_Particles.exe with the users values and write the output to the reults file with the prefix Detected_
system(cmd)

%Show detection results - Red Original Image -ROIs->White -
% Green/Yellow->Excluded by filters
figure('Name','Detected Particles - Red Original Image - Blue to White Selected ROIs - Green to Yellow->Excluded by filters')
channel1Im = rescale(imread([workingDir,'Aligned_Partial_Mean.tiff']),displayMin,displayMax);
channel1Im=min(max(channel1Im,0),1);
channel2Im = rescale(imread([workingDir,'Detected_Regions.tif']));
channel3Im = rescale(imread([workingDir,'Detected_Filtered_Regions.tif']));
combinedImage = cat(3, overlayColour1(1).*channel1Im+overlayColour2(1).*channel2Im+overlayColour3(1).*channel3Im,overlayColour1(2).*channel1Im+overlayColour2(2).*channel2Im+overlayColour3(2).*channel3Im,overlayColour1(3).*channel1Im+overlayColour2(3).*channel2Im+overlayColour3(3).*channel3Im);
imshow(combinedImage)
truesize([900 900]);
disp('Finish detecting particles');

%% 5) Expand Regions
foregroundDist = 4.1; % Distance to dilate the ROIs by to make sure all flourescence from the ROI is measured
backInnerDist = 4.1;
backOuterDist = 20; % Distance to dilate beyond the ROI to measure the local background

displayMin = 0; % This just adjusts the contrast in the displayed image. It does NOT effect detection
displayMax = 2; % This just adjusts the contrast in the displayed image. It does NOT effect detection

cmd = [JIM,'Expand_Shapes.exe "',workingDir,'Detected_Filtered_Positions.csv" "',workingDir,'Detected_Positions.csv" "',workingDir,'Expanded" -boundaryDist ', num2str(foregroundDist),' -backgroundDist ',num2str(backOuterDist),' -backInnerRadius ',num2str(backInnerDist)]; % Run Fit_Arbitrary_Shapes.exe on the Detected_Filtered_Positions and output the result with the prefix Expanded
system(cmd)

%show expansion reult
figure('Name','Detected Particles - Red Original Image - Green ROIs - Blue Background Regions')
channel1Im = rescale(imread([workingDir,'Aligned_Partial_Mean.tiff']),displayMin,displayMax);
channel1Im=min(max(channel1Im,0),1);
channel2Im = rescale(imread([workingDir,'Expanded_Channel_ROIs.tif']));
channel3Im = rescale(imread([workingDir,'Expanded_Channel_Background_Regions.tif']));
combinedImage = cat(3, overlayColour1(1).*channel1Im+overlayColour2(1).*channel2Im+overlayColour3(1).*channel3Im,overlayColour1(2).*channel1Im+overlayColour2(2).*channel2Im+overlayColour3(2).*channel3Im,overlayColour1(3).*channel1Im+overlayColour2(3).*channel2Im+overlayColour3(3).*channel3Im);
imshow(combinedImage);
truesize([900 900]);

%% 6) Calculate Traces
verboseOutput = true;

verboseString = '';
if verboseOutput
    verboseString = ' -Verbose';
end

cmd = [JIM,'Calculate_Traces.exe "',completeName,'" "',workingDir,'Expanded_ROI_Positions.csv" "',workingDir,'Expanded_Background_Positions.csv" "',workingDir,'Channel_1" -Drift "',workingDir,'Aligned_Drifts.csv"',verboseString]; % Generate traces using AS_Measure_Each_Frame.exe and write out with the prefix Channel_1
system(cmd)

variableString = ['Date, ', datestr(datetime('today')),'\n'...
    ,'iterations,',num2str(iterations),'\nalignStartFrame,', num2str(alignStartFrame),'\nalignEndFrame,', num2str(alignEndFrame),'\n'...
    ,'useMaxProjection,',num2str(useMaxProjection),'\ndetectionStartFrame,', num2str(detectionStartFrame),'\ndetectionEndFrame,', num2str(detectionEndFrame),'\n'...
    ,'cutoff,',num2str(cutoff),'\nleft,', num2str(left),'\nright,', num2str(right),'\ntop,', num2str(top),'\nbottom,', num2str(bottom),'\n'...
    ,'minCount,',num2str(minCount),'\nmaxCount,', num2str(maxCount),'\nminEccentricity,', num2str(minEccentricity),'\nmaxEccentricity,', num2str(maxEccentricity),'\n'...
    ,'minLength,',num2str(minLength),'\nmaxLength,', num2str(maxLength),'\nmaxDistFromLinear,', num2str(maxDistFromLinear),'\n'...
    ,'foregroundDist,',num2str(foregroundDist),'\nbackInnerDist,', num2str(backInnerDist),'\nbackOuterDist,', num2str(backOuterDist),'\nverboseOutput,', num2str(verboseOutput)];

fileID = fopen([workingDir,'Trace_Generation_Variables.csv'],'w');
fprintf(fileID, variableString);
fclose(fileID);

%% 7) View Traces
pageNumber = 3;

traces=csvread([workingDir,'\Channel_1_Fluorescent_Intensities.csv'],1);
measures = csvread([workingDir,'\Detected_Filtered_Measurements.csv'],1);
channel1Im = imread([workingDir,'Detected_Filtered_Region_Numbers.tif']);
figure('Name','Particle Numbers');
imshow(channel1Im);
truesize([900 900]);

figure
set(gcf, 'Position', [100, 100, 1500, 800])

for i=1:36
    if i+36*(pageNumber-1)<size(traces,1)
    subplot(6,6,i)
    hold on
    title(['Particle ' num2str(i+36*(pageNumber-1)) ' x ' num2str(round(measures(i+36*(pageNumber-1),1))) ' y ' num2str(round(measures(i+36*(pageNumber-1),2)))])
    plot(traces(i+36*(pageNumber-1),:),'-r');
    plot([0 size(traces(i+36*(pageNumber-1),:),2)],[0 0] ,'-b');
    xlim([0 size(traces(i+36*(pageNumber-1),:),2)])
    hold off
    end
end

%% Continue from here for batch processing
%
%
%
%
%
%% 8) Detect files for batch
filesInSubFolders = false; % Set this to true if each image stack is in it's own folder or false if imagestacks are directly in the main folder




[jimPath,~,~] = fileparts(matlab.desktop.editor.getActivefileName); % Find the location of this script again in case the user is just running batch (should be in Jim\Matlab_Programs)
JIM = [fileparts(jimPath),'\Jim_Programs\'];
fileName = uigetdir(); % open the dialog box to select the folder for batch files
fileName=[fileName,'\'];

if filesInSubFolders
    allFiles = dir(fileName); % find everything in the input folder
    allFiles(~[allFiles.isdir]) = []; % filter for folders
    allFiles=allFiles(3:end);
    allFilescells = arrayfun(@(y) arrayfun(@(x) [fileName,y.name,'\',x.name],[dir([fileName,y.name,'\*.tif']); dir([fileName,y.name,'\*.tiff'])]','UniformOutput',false),allFiles','UniformOutput',false); % look in each folder and pull out all files that end in tif or tiff
    allFilescells = horzcat(allFilescells{:})'; % combine the files from all folders into one list
    NumberOfFiles=size(allFilescells,1);
else
    allFiles = [dir([fileName,'\*.tif']); dir([fileName,'\*.tiff'])];% find everything in the main folder ending in tiff or tif
    allFilescells = arrayfun(@(y) [fileName,y.name],allFiles,'UniformOutput',false); % generate a full path name for each file
    allFilescells = allFilescells(~startsWith(allFilescells,[fileName,'.']));
    NumberOfFiles=size(allFilescells,1);
end
disp(['There are ',num2str(NumberOfFiles),' files to analyse']);

%% 9) Batch Analyse
overwritePreviouslyAnalysed = true;
parfor i=1:NumberOfFiles(1)
    completeName = allFilescells{i};
    disp(['Analysing ',completeName]);
    % 3.2) Create folder for results
    [fileNamein,name,~] = fileparts(completeName);%get the name of the tiff image
    workingDir = [fileNamein,'\',name];
    [fileNamein,name,~] = fileparts(workingDir);
    workingDir = [fileNamein,'\',name,'\'];
    if ~exist(workingdir, 'dir')
        mkdir(workingdir)%make a subfolder with that name
    end

    if (exist([workingDir,'Channel_1_Fluorescent_Intensities.csv'],'file')==2 && overwritePreviouslyAnalysed==false)
        disp(['Skipping ',completeName,' - Analysis already exists']);
        continue
    end
    
    % 3.3)  Calculate Drifts
    cmd = [JIM,'Align_Channels.exe "',workingDir,'Aligned" "',completeName,'" -Start ',num2str(alignStartFrame),' -End ',num2str(alignEndFrame),' -Iterations ',num2str(iterations)];%Run the Align_Channels program with the selected image stack as the input and save the results to the results folder with the Aligned prefix
    system(cmd);

    
    cmd = [JIM,'Mean_of_Frames.exe NULL "',workingDir,'Aligned_Drifts.csv" "',workingDir,'Aligned" "',completeName,'" -Start ',num2str(detectionStartFrame),' -End ',num2str(detectionEndFrame),maxProjectionString];
    system(cmd);
    % 3.4) Detect Particles
    cmd = [JIM,'Detect_Particles.exe "',workingDir,'Aligned_Partial_Mean.tiff" "',workingDir,'Detected" -BinarizeCutoff ', num2str(cutoff),' -minLength ',num2str(minLength),' -maxLength ',num2str(maxLength),' -minCount ',num2str(minCount),' -maxCount ',num2str(maxCount),' -minEccentricity ',num2str(minEccentricity),' -maxEccentricity ',num2str(maxEccentricity),' -left ',num2str(left),' -right ',num2str(right),' -top ',num2str(top),' -bottom ',num2str(bottom),' -maxDistFromLinear ',num2str(maxDistFromLinear)]; % Run the program Find_Particles.exe with the users values and write the output to the reults file with the prefix Detected_
    system(cmd)
    % 3.5) Fit areas around each shape
    cmd = [JIM,'Expand_Shapes.exe "',workingDir,'Detected_Filtered_Positions.csv" "',workingDir,'Detected_Positions.csv" "',workingDir,'Expanded" -boundaryDist ', num2str(foregroundDist),' -backgroundDist ',num2str(backOuterDist),' -backInnerRadius ',num2str(backInnerDist)]; % Run Fit_Arbitrary_Shapes.exe on the Detected_Filtered_Positions and output the result with the prefix Expanded
    system(cmd)
    % 3.6) Calculate Sum of signal and background for each frame
    cmd = [JIM,'Calculate_Traces.exe "',completeName,'" "',workingDir,'Expanded_ROI_Positions.csv" "',workingDir,'Expanded_Background_Positions.csv" "',workingDir,'Channel_1" -Drifts "',workingDir,'Aligned_Drifts.csv"',verboseString]; % Generate traces using AS_Measure_Each_Frame.exe and write out with the prefix Channel_1
    system(cmd)
    
    fileID = fopen([workingDir,'Trace_Generation_Variables.csv'],'w');
    fprintf(fileID, variableString);
    fclose(fileID);

end

File: Jim_v5_7/Python_Programs/Generate_Single_Channel_Traces.py
import datetime
from sys import platform as _platform
import sys
import os
import tkinter as tk
import matplotlib.pyplot as plt
import matplotlib.image as mpimg
import csv
import numpy as np
from tkinter import filedialog
from PIL import Image



sectionNumber = 7
# Sections
# 1 - Select input file and create a folder for results
# 2 - Drift Correct
# 3 - Make a SubAverage of Frames for Detection
# 4 - Detect Particles
# 5 - Expand Regions
# 6 - Calculate Traces
# 7 - View Traces
# 8 - Detect files for batch
# 9 - Batch Analyse

# ~~~~PARAMETER INPUTS~~~~ #
# 1 - General Parameters
overlayColours1 = [1, 0, 0]
overlayColours2 = [0, 1, 0]
overlayColours3 = [0, 0, 1]

displayMin = 0
displayMax = 2


# 2 - Drift Correct Parameters
iterations = 3
alignStartFrame = 1
alignEndFrame = 5

# 3 - Make a SubAverage of the Image Stack for Detection Parameters
useMaxProjection = False
detectionStartFrame = 1
detectionEndFrame = 25

# 4 - Detect Particles Parameters
cutoff = 0.85  # The cutoff for the initial threshold

# Filtering
left = 10  # Excludes particles closer to the left edge than this (in pixels).
right = 10  # Excludes particles closer to the right edge than this (in pixels). 
top = 10  # Excludes particles closer to the top edge than this (in pixels). 
bottom = 10  # Excluded particles closer to the bottom edge than this (in pixels). 


minCount = 10  # Minimum number of pixels in a region to be included
maxCount = 100  # Maximum number of pixels in a region to be included

# Eccentricity of best fit ellipse goes from 0 to 1 - 0=Perfect Circle, 1 = Line.
minEccentricity = -0.1  # Use the Minimum to exclude round objects. Set below zero to allow all round objects
maxEccentricity = 1.1  # Use the Maximum to exclude long, thin objects. Set  above 1 to include all long, thin objects

minLength = 0  # Minimum length of the region
maxLength = 100000  # Maximum length of the region

maxDistFromLinear = 10000000  # Maximum distance that a pixel can deviate from the major axis.

# 5 - Expand Regions Parameters

foregroundDist = 4.1  # Distance to dilate the ROIs by to make sure all flourescence from the ROI is measured
backInnerDist = 4.1
backOuterDist = 20  # Distance to dilate beyond the ROI to measure the local background

# 6 - Calculate Traces Parameter
verboseOutput = False

# 7 - View Traces Parameter
pageNumber = 1

# 8 - Detect files for batch
filesInSubFolders = False

# 9 - Batch Analysis
overwritePreviouslyAnalysed = True


# ~~~~ Don't touch from here ~~~~ #
pyfile = sys.argv[0]
if _platform == "linux" or _platform == "linux2":
    JIM = os.path.dirname(os.path.dirname(pyfile)) + "/Jim_Programs_Linux/"
    fileSep = '/'
    fileEXE = ''
elif _platform == "darwin":
    JIM = os.path.dirname(os.path.dirname(pyfile)) + "/Jim_Programs_Mac/"
    fileSep = '/'
    fileEXE = ''
elif _platform == "win32" or _platform == "win64":
    JIM = os.path.dirname(os.path.dirname(pyfile)) + "\\Jim_Programs\\"
    fileEXE = '.exe'


#  Change if not running in original distribution folder e.g. JIM = "C:\\Users\\James\\Desktop\\Jim_v5\\Jim_Programs\\"

if sectionNumber != 1 and sectionNumber != 8:
    channelFile = os.path.dirname(pyfile) + fileSep + "saveFilename.csv"
    fileList = list(csv.reader(open(channelFile)))
    completeName = fileList[0][0]
    print("Analysing "+completeName)
    workingDir = os.path.dirname(completeName) + fileSep + str(os.path.basename(completeName).split(".", 1)[0]) + fileSep


# 1 - Select input file and create a folder for results
if sectionNumber == 1:
    root = tk.Tk()
    root.withdraw()

    completeName = filedialog.askopenfilename()
    completeName = completeName.replace("/", fileSep)
    completeName = completeName.replace("\\", fileSep)
    print(completeName)
    workingDir = os.path.dirname(completeName)+fileSep+str(os.path.basename(completeName).split(".", 1)[0])+fileSep
    print(workingDir)
    if not os.path.isdir(workingDir):
        os.makedirs(workingDir)
    savedFilenameFile = open(os.path.dirname(pyfile) + fileSep + "saveFilename.csv", "w")
    savedFilenameFile.write(completeName)
    savedFilenameFile.close()

# 2 - Drift Correct
if sectionNumber == 2:

    cmd = (JIM + 'Align_Channels' + fileEXE + ' \"' + workingDir + "Aligned\" \"" + completeName + "\" -Start "
           + str(alignStartFrame) + " -End " + str(alignEndFrame) + ' -Iterations '+str(iterations))
    os.system(cmd)

    imName = workingDir+"Aligned_initial_mean.tiff"
    img = mpimg.imread(imName)
    plt.figure('Before Drift Correction')
    plt.imshow(img, cmap="gray")

    imName = workingDir+"Aligned_final_mean.tiff"
    img = mpimg.imread(imName)
    plt.figure('After Drift Correction')
    plt.imshow(img, cmap="gray")
    plt.show()

# 3 - Make a SubAverage of the Image Stack for Detection
if sectionNumber == 3:

    maxProjectionString = ""
    if useMaxProjection:
        maxProjectionString = " -MaxProjection"

    cmd = (JIM + 'Mean_of_Frames' + fileEXE + ' NULL \"' + workingDir + "Aligned_Drifts.csv\" \""
           + workingDir + "Aligned\" \"" + completeName + "\" -Start " + str(detectionStartFrame)
           + " -End " + str(detectionEndFrame) + maxProjectionString)
    os.system(cmd)

    imName = workingDir+"Aligned_Partial_Mean.tiff"
    img = mpimg.imread(imName)
    plt.figure('Sub-Average to use for detection')
    imPlot = plt.imshow(img, cmap="gray")
    plt.show()


# 4 - detect particles
if sectionNumber == 4:

    cmd = (JIM + 'Detect_Particles' + fileEXE + ' \"' + workingDir + 'Aligned_Partial_Mean.tiff\" \"' + workingDir
           + 'Detected\" -BinarizeCutoff ' + str(cutoff) + ' -minLength ' + str(minLength) + ' -maxLength '
           + str(maxLength) + ' -minCount ' + str(minCount) + ' -maxCount ' + str(maxCount)
           + ' -minEccentricity ' + str(minEccentricity)+ ' -maxEccentricity ' + str(maxEccentricity)
           + ' -maxDistFromLinear ' + str(maxDistFromLinear)
           + ' -left ' + str(left) + ' -right ' + str(right) + ' -top ' + str(top) + ' -bottom ' + str(bottom))
    os.system(cmd)

    channel1Im = np.array(Image.open(workingDir+"Aligned_Partial_Mean.tiff")).astype(float)
    channel1Im = np.clip(255*(displayMax-displayMin)*(channel1Im - np.min(channel1Im)) / np.ptp(channel1Im)-displayMin, 0, 255)
    channel2Im = np.array(Image.open(workingDir + "Detected_Regions.tif"))
    channel3Im = np.array(Image.open(workingDir + "Detected_Filtered_Regions.tif"))
    combinedImage = (np.dstack((overlayColours1[0]*channel1Im + overlayColours2[0]*channel2Im + overlayColours3[0]*channel3Im,
                                overlayColours1[1]*channel1Im + overlayColours2[1]*channel2Im + overlayColours3[1]*channel3Im,
                                overlayColours1[2]*channel1Im + overlayColours2[2]*channel2Im + overlayColours3[2]*channel3Im))).astype(np.uint8)
    imPlot = plt.imshow(combinedImage)
    plt.show()


# 5 - Expand Regions
if sectionNumber == 5:
    cmd = (JIM + 'Expand_Shapes' + fileEXE + ' \"' + workingDir + 'Detected_Filtered_Positions.csv\" \"'
           + workingDir + 'Detected_Positions.csv\" \"' + workingDir + 'Expanded\" -boundaryDist '
           + str(foregroundDist) + ' -backgroundDist ' + str(backOuterDist) + ' -backInnerRadius '+str(backInnerDist))
    os.system(cmd)

    channel1Im = np.array(Image.open(workingDir+"Aligned_Partial_Mean.tiff")).astype(float)
    channel1Im = np.clip(255*(displayMax-displayMin)*(channel1Im - np.min(channel1Im)) / np.ptp(channel1Im)-displayMin, 0, 255)
    channel2Im = np.array(Image.open(workingDir + "Expanded_ROIs.tif"))
    channel3Im = np.array(Image.open(workingDir + "Expanded_Background_Regions.tif"))
    combinedImage = (np.dstack((overlayColours1[0]*channel1Im + overlayColours2[0]*channel2Im + overlayColours3[0]*channel3Im,
                                overlayColours1[1]*channel1Im + overlayColours2[1]*channel2Im + overlayColours3[1]*channel3Im,
                                overlayColours1[2]*channel1Im + overlayColours2[2]*channel2Im + overlayColours3[2]*channel3Im))).astype(np.uint8)
    imPlot = plt.imshow(combinedImage)
    plt.show()

# 6 - Calculate Traces
if sectionNumber == 6:
    verboseString = ''
    if verboseOutput:
        verboseString = ' -Verbose'

    cmd = (JIM + 'Calculate_Traces' + fileEXE + ' \"' + completeName + '\" \"' + workingDir + 'Expanded_ROI_Positions.csv\" \"'
           + workingDir + 'Expanded_Background_Positions.csv\" \"' + workingDir + 'Channel_1\" -Drift \"'
           + workingDir + 'Aligned_Drifts.csv\"'+verboseString)
    os.system(cmd)

    variableString = ('Date, ' + str(datetime.date.today()) + '\niterations,' + str(iterations) +
                      '\nalignStartFrame,' + str(alignStartFrame) + '\nalignEndFrame,' + str(alignEndFrame) +
                      '\nuseMaxProjection,' + str(int(useMaxProjection)) + '\ndetectionStartFrame,' + str(detectionStartFrame) +
                      '\ndetectionEndFrame,' + str(detectionEndFrame) + '\ncutoff,' + str(cutoff) +
                      '\nleft,' + str(left) + '\nright,' + str(right) + '\ntop,' + str(top) + '\nbottom,' + str(bottom) +
                      '\nminCount,' + str(minCount) + '\nmaxCount,' + str(maxCount) +
                      '\nminEccentricity,' +str(minEccentricity) + '\nmaxEccentricity,' + str(maxEccentricity) +
                      '\nminLength,' + str(minLength) + '\nmaxLength,' + str(maxLength) +
                      '\nmaxDistFromLinear,' + str(maxDistFromLinear) + '\nforegroundDist,' + str(foregroundDist) +
                      '\nbackInnerDist,' + str(backInnerDist) + '\nbackOuterDist,' + str(backOuterDist) +
                      '\nverboseOutput,' + str(int(verboseOutput)))

    saveVariablesFile = open(workingDir + "Trace_Generation_Variables.csv", "w")
    saveVariablesFile.write(variableString)
    saveVariablesFile.close()

# 7 - View Traces
if sectionNumber == 7:
    imName = workingDir + "Detected_Filtered_Region_Numbers.tif"
    img = mpimg.imread(imName)
    plt.figure('Before Drift Correction')
    imPlot = plt.imshow(img, cmap="gray")

    channelFile = workingDir+"Detected_Filtered_Measurements.csv"
    measurements = list(csv.reader(open(channelFile)))

    channelFile = workingDir+"Channel_1_Fluorescent_Intensities.csv"
    data = list(csv.reader(open(channelFile)))
    plt.figure(figsize=(13, 8))
    for i in range(1, 37):
        if len(data) > i+(pageNumber-1)*36:
            plt.subplot(6, 6, i)
            plt.plot([float(i) for i in data[i+(pageNumber-1)*36]])
            plt.plot(plt.xlim(), [0, 0], color='black')
            xpos = round(float(measurements[i+(pageNumber-1)*36][0]))
            ypos = round(float(measurements[i+(pageNumber-1)*36][1]))
            plt.title('Particle '+str(i+(pageNumber-1)*36)+' x '+str(xpos)+' y '+str(ypos))
    plt.tight_layout()
    plt.show()

# 8 - Detect files for batch
if sectionNumber == 8:
    root = tk.Tk()
    root.withdraw()
    topFolder = filedialog.askdirectory(parent=root)

    fileList = []
    if filesInSubFolders:
        for folder in os.listdir(topFolder):
            folderin = os.path.join(topFolder, folder)
            if os.path.isdir(folderin):
                for file in os.listdir(folderin):
                    if file.endswith(".tif") or file.endswith(".tiff") or file.endswith(".TIF") or file.endswith(".TIFF"):
                        fileList.append(os.path.join(folderin, file).replace("/", fileSep))
    else:
        for file in os.listdir(topFolder):
            if file.endswith(".tif") or file.endswith(".tiff") or file.endswith(".TIF") or file.endswith(".TIFF"):
                fileList.append(os.path.join(topFolder, file).replace("/", fileSep))

    print('There are ' + str(len(fileList))+' files to analyse')
    print(fileList)
    savedFilenameFile = open(os.path.dirname(pyfile) + fileSep + "saveFilename.csv", "w")
    for i in range(len(fileList)):
        savedFilenameFile.write(fileList[i]+"\n")
    savedFilenameFile.close()


if sectionNumber == 9:

    maxProjectionString = ""
    if useMaxProjection:
        maxProjectionString = " -MaxProjection"

    for filein in fileList:
        completeName = filein[0]
        print('Analysing file '+completeName)
        workingDir = os.path.dirname(completeName) + fileSep + str(os.path.basename(completeName).split(".", 1)[0]) + fileSep

        if os.path.isfile(workingDir + "Channel_1_Fluorescent_Intensities.csv"):
            print('Skipping ' + completeName + ' - Analysis already exists')
            continue

        if not os.path.isdir(workingDir):
            os.makedirs(workingDir)

        cmd = (JIM + 'Align_Channels' + fileEXE + ' \"' + workingDir + "Aligned\" \"" + completeName + "\" -Start "
               + str(alignStartFrame) + " -End " + str(alignEndFrame) + ' -Iterations ' + str(iterations))
        os.system(cmd)

        cmd = (JIM + 'Mean_of_Frames' + fileEXE + ' NULL \"' + workingDir + "Aligned_Drifts.csv\" \"" + workingDir + "Aligned\" \""
               + completeName + "\" -Start " + str(detectionStartFrame)
               + " -End " + str(detectionEndFrame) + maxProjectionString)
        os.system(cmd)

        cmd = (JIM + 'Detect_Particles' + fileEXE + ' \"' + workingDir + 'Aligned_Partial_Mean.tiff\" \"' + workingDir
               + 'Detected\" -BinarizeCutoff ' + str(cutoff) + ' -minLength ' + str(minLength) + ' -maxLength '
               + str(maxLength) + ' -minCount ' + str(minCount) + ' -maxCount ' + str(maxCount)
               + ' -minEccentricity ' + str(minEccentricity) + ' -maxEccentricity ' + str(maxEccentricity)
               + ' -maxDistFromLinear ' + str(maxDistFromLinear)
               + ' -left ' + str(left) + ' -right ' + str(right) + ' -top ' + str(top) + ' -bottom ' + str(bottom))
        os.system(cmd)

        cmd = (JIM + 'Expand_Shapes' + fileEXE + ' \"' + workingDir + 'Detected_Filtered_Positions.csv\" \"'
               + workingDir + 'Detected_Positions.csv\" \"' + workingDir + 'Expanded\" -boundaryDist '
               + str(foregroundDist) + ' -backgroundDist ' + str(backOuterDist) + ' -backInnerRadius ' + str(
                    backInnerDist))
        os.system(cmd)

        cmd = (JIM + 'Calculate_Traces' + fileEXE + ' \"' + completeName + '\" \"' + workingDir + 'Expanded_ROI_Positions.csv\" \"'
               + workingDir + 'Expanded_Background_Positions.csv\" \"' + workingDir + 'Channel_1\" -Drift \"'
               + workingDir + 'Aligned_Drifts.csv\"')
        os.system(cmd)

        variableString = ('Date, ' + str(datetime.date.today()) + '\niterations,' + str(iterations) +
                          '\nalignStartFrame,' + str(alignStartFrame) + '\nalignEndFrame,' + str(alignEndFrame) +
                          '\nuseMaxProjection,' + str(int(useMaxProjection)) + '\ndetectionStartFrame,' + str(detectionStartFrame) +
                          '\ndetectionEndFrame,' + str(detectionEndFrame) + '\ncutoff,' + str(cutoff) +
                          '\nleft,' + str(left) + '\nright,' + str(right) + '\ntop,' + str(top) + '\nbottom,' + str(bottom) +
                          '\nminCount,' + str(minCount) + '\nmaxCount,' + str(maxCount) +
                          '\nminEccentricity,' + str(minEccentricity) + '\nmaxEccentricity,' + str(maxEccentricity) +
                          '\nminLength,' + str(minLength) + '\nmaxLength,' + str(maxLength) +
                          '\nmaxDistFromLinear,' + str(maxDistFromLinear) + '\nforegroundDist,' + str(foregroundDist) +
                          '\nbackInnerDist,' + str(backInnerDist) + '\nbackOuterDist,' + str(backOuterDist) +
                          '\nverboseOutput,' + str(int(verboseOutput)))

        saveVariablesFile = open(workingDir + fileSep + "Trace_Generation_Variables.csv", "w")
        saveVariablesFile.write(variableString)
        saveVariablesFile.close()

    print('Batch Analysis Completed')
